app/models/job_assignment_log.rb

```diff
@@ -0,0 +1,62 @@
+class JobAssignmentLog < ActiveRecord::Base
+
+  attr_accessible :job_id, :assigned_by, :assigned_to
+
+  belongs_to :job
+  belongs_to :assigned_by_user, class_name: "User", foreign_key: "assigned_by"
+  belongs_to :assigned_to_user, class_name: "User", foreign_key: "assigned_to"
+
+  ### VALIDATIONS
+
+  # VALIDATE FOREIGN KEYS ON SAVE
+  def save
+    err = nil
```

Comment:
I disagree. In general I would prefer to let the user know they have 3 errors on a form than have them fix one, try again, fix another, try again, then fix another...  Imagine if you were filling out a complex form and were given errors one at a time. I think it is cleaner to show all <n> inputs that you need to fix all at once!  

That said, for this particular example our UX actually gates this differently upstream:
- the controller stops the user if the job_id is invalid (regardless of the assigned_by and assigned_to inputs).
- the "assigned by" is the currently logged-in user, so it has to be legit
- the "assigned to" comes from a dropdown. It could be wrong if there is an edge case, but should not happen in "normal" use.

With all of that in mind  I would prefer that the model check everything and report back on everything (for extensibility, etc), even if there are other gates upstream in the system.  But I am happy to discuss and change it.

Reply:
If the side-effects are a major part of what is happening, lift them up so that they aren't hidden.  Maybe like this

```ruby
def validate
  valid = true
  unless Job.exists?(self.job_id)
    self.errors[:job_id] << 'invalid job'
    valid = false
  end
  unless User.exists?(self.assigned_by)
    self.errors[:by_id] << "invalid user"
    valid = false
  end
  unless ...
    ...
  end

  valid
end
```
Minimally the method should have a block comment that points out the side-effects